File: deck/base/code/native/rust/process/runtime/runner.rs
// Subprocess runner over std::process::Command. Runs the command to completion, capturing stdout and stderr, and
// returns the exit code with both streams. A spawn failure returns code -1 and the error text, so the public run API
// stays total. Reached only through the public run API.
mod runner {
    use super::RunResult;
    pub async fn run(command: String, argument_list: Vec<String>) -> RunResult {
        match std::process::Command::new(&command).args(&argument_list).output() {
            Ok(out) => RunResult {
                code: out.status.code().unwrap_or(0) as i64,
                output: String::from_utf8_lossy(&out.stdout).to_string(),
                error: String::from_utf8_lossy(&out.stderr).to_string(),
            },
            Err(cause) => RunResult {
                code: -1,
                output: String::new(),
                error: cause.to_string(),
            },
        }
    }
}

File: deck/base/code/native/node/process/runtime/runner.ts
// Subprocess runner over node:child_process. Spawns the command, accumulates stdout and stderr, and resolves with the
// exit code and captured streams when the process closes. A spawn failure resolves with code -1 and the error text, so
// the public run API stays total. Reached only through the public run API.
import { spawn } from 'node:child_process'

const runner = {
  run: (
    command: string,
    argumentList: string[],
  ): Promise<{ code: number; output: string; error: string }> =>
    new Promise(resolve => {
      let output = ''
      let error = ''
      try {
        const child = spawn(command, argumentList)
        child.stdout.on('data', (chunk: Buffer) => {
          output += chunk.toString()
        })
        child.stderr.on('data', (chunk: Buffer) => {
          error += chunk.toString()
        })
        child.on('error', (cause: Error) => {
          resolve({ code: -1, output, error: error + String(cause) })
        })
        child.on('close', (code: number | null) => {
          resolve({ code: code ?? 0, output, error })
        })
      } catch (cause) {
        resolve({ code: -1, output, error: String(cause) })
      }
    }),
}
